ios/AudioInputController.m

```diff
@@ -10,61 +10,76 @@
 @implementation AudioInputController {
     AudioComponentInstance remoteIOUnit;
     BOOL isInitialized;
-}
-
-- (void)dealloc {
-    NSLog(@"[AudioInputController][dealloc]");
-    AudioComponentInstanceDispose(remoteIOUnit);
+    NSString* origAudioCategory;
+    NSString* origAudioMode;
 }
 
 + (instancetype) sharedInstance {
     static AudioInputController *instance = nil;
     static dispatch_once_t onceToken;
-    
+
     dispatch_once(&onceToken, ^{
         instance = [[self alloc] init];
     });
-    
+
     return instance;
 }
 
 - (OSStatus) start {
-    return AudioOutputUnitStart(self->remoteIOUnit);
+    return AudioOutputUnitStart(remoteIOUnit);
 }
 
 - (OSStatus) stop {
-    return AudioOutputUnitStart(self->remoteIOUnit);
+    OSStatus res = AudioOutputUnitStop(remoteIOUnit);
+    AudioComponentInstanceDispose(remoteIOUnit);
+    remoteIOUnit = nil;
+    isInitialized = NO;
+
+    [self restoreOriginalAudioSetup];
+    [[AVAudioSession sharedInstance] setActive:NO withOptions:AVAudioSessionSetActiveOptionNotifyOthersOnDeactivation error:nil];
+    return res;
 }
 
 - (OSStatus) prepareWithSampleRate:(double)desiredSampleRate {
     NSLog(@"[WebRTCVad] sampleRate = %f", desiredSampleRate);
-    
-    AVAudioSession *session = [AVAudioSession sharedInstance];
-    double sampleRate = session.sampleRate;
-    
-    NSLog(@"hardware sample rate = %f, using specified rate = %f", sampleRate, desiredSampleRate);
-    
+
+    AVAudioSession *audioSession = [AVAudioSession sharedInstance];
+    double sampleRate = audioSession.sampleRate;
+
+    NSLog(@"[WebRTCVad] hardware sample rate = %f, using specified rate = %f", sampleRate, desiredSampleRate);
+
     if (desiredSampleRate){
         sampleRate = desiredSampleRate;
     }
-    
+
     // Store audio sample rate for later use
     self.audioSampleRate = sampleRate;
-    
-    NSError *error;
-    BOOL ok = [session setCategory:AVAudioSessionCategoryPlayAndRecord error:&error];
-    
-    NSLog(@"set category %d", ok);
-    
-    // 5 second audio buffer hint
-    [session setPreferredIOBufferDuration:5 error:&error];
-    
+
+    @try {
+        [self storeOriginalAudioSetup];
+
+        BOOL ok = [audioSession setCategory:AVAudioSessionCategoryPlayAndRecord error:nil];
+
+        NSLog(@"[WebRTCVad] set category %d", ok);
+
+        [audioSession setMode:AVAudioSessionModeVoiceChat error:nil];
+
+        NSLog(@"[WebRTCVad] set mode %d", ok);
+
+        // 5 second audio buffer hint
+        [audioSession setPreferredIOBufferDuration:5 error:nil];
+
+    } @catch (NSException *e) {
+        NSLog(@"[WebRTCVad]: session setup failed: %@", e.reason);
```

Comment:
would it be worthwhile to pass in an ErrorLogger of some sort to this module such that we can get more sophisticated logging (i.e., Sentry), or is this exception mostly noise due to permissions?

Reply:
it seems like this module could at least benefit from some common logger that will append this prefix consistently

Reply:
Yeah, I agree, at least a common function. The tricky thing here, though, is we need to specify formatters for each arg, which changes based on arg type. ie - %d -> digit, %f -> decimal, float, %@ -> object, etc. Will think on this a bit and consolidate